src/main/java/edu/harvard/iq/dataverse/externaltools/ExternalToolHandler.java

```diff
@@ -97,9 +101,22 @@ private String getQueryParam(String key, String value) {
         if (hardCodedKludgeForDataExplorer) {
             if (key.equals("uri")) {
                 String staticSiteUrl = SystemConfig.getDataverseSiteUrlStatic();
-                // TODO: support {siteUrl} for real instead of having replaceAll in here. Write a parser.
-                // The "\\" is so replaceAll doesn't freak out at "{" and throw"java.util.regex.PatternSyntaxException: Illegal repetition"
-                return key + "=" + value.replaceAll("\\" + ReservedWord.SITE_URL, staticSiteUrl).replaceAll("\\" + ReservedWord.FILE_ID, getDataFile().getId().toString());
+                //create a map for special word replacement
+                Map<String, String> replacements = new HashMap<String, String>() {{
+                    put("siteUrl", staticSiteUrl);
+                    put("fileId", getDataFile().getId().toString());
+                }};
+                StringBuffer sb = new StringBuffer();
+                Pattern p = Pattern.compile("\\{(.*?)\\}");
+                Matcher m = p.matcher(value);
+                while (m.find()) {
+                    String repString = replacements.get(m.group(1));
+                    if (repString != null){   
+                        m.appendReplacement(sb, repString);
+                    }
+                }
+
+                return key + "=" + m.appendTail(sb);
```

Comment:
The above is an interesting approach. It's new to me. One thing I'd like to mention is that I wrote together both `getQueryParametersForUrl` and `getQueryParam` but I don't mind if all their code is completely rewritten. I wrote a variety of tests to assert that at the end the query parameters come out the way we want them but, again, I don't mind if the implementation is rewritten.